src/lib.rs

```diff
@@ -151,6 +161,45 @@ impl<I: DoubleEndedIterator> DoubleEndedIterator for PushBackIterator<I> {
     }
 }
 
+/// A lookahead iterator that doesn't consume the borrowed [`PushBackIterator`].
+///
+/// Unfortunately, we have to require [`Clone`] for the item type, because the
+/// [`Iterator`] trait, as it is defined in the standard library, does not
+/// allow us to return a reference that mutably borrows the iterator itself.
+///
+/// This borrow is needed because it is not safe to advance the iterator when
+/// the previous returned item is still alive, as this item is borrowing the
+/// VecDeque that we have to mutate.
+///
+pub struct LookaheadIterator<'i, I: Iterator>
+{
+    inner: &'i mut PushBackIterator<I>,
+    pos: usize,
+}
+
+impl <'i, I: Iterator> Iterator for LookaheadIterator<'i, I>
+    where I::Item: Clone
+{
+    type Item = I::Item;
+
+    fn next(self: &mut LookaheadIterator<'i, I>) -> Option<I::Item> {
+        self.inner.peek_nth(self.pos).cloned()
```

Comment:
I don't fully understand what this code is doing. Is this suppose to be incrementing `self.pos`?

Reply:
Yes, I must not have had enough coffee :o